[azservicebus] Data race on link.err when closing connection

### Bug Report

This race is detected by the race detector when running the tests in https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/messaging/azservicebus with `go test -race -v`:

```
=== RUN   TestAdminClient_UsingIdentity
    admin_client_test.go:31: Azure Identity compatible credentials not configured
--- SKIP: TestAdminClient_UsingIdentity (0.00s)
=== RUN   TestAdminClient_QueueWithMaxValues
--- PASS: TestAdminClient_QueueWithMaxValues (3.41s)
=== RUN   TestAdminClient_AddQueue
--- PASS: TestAdminClient_AddQueue (1.63s)
=== RUN   TestAdminClient_Queue_Forwarding
==================
WARNING: DATA RACE
Read at 0x00c00045a0c0 by goroutine 37:
  github.com/Azure/go-amqp.(*Receiver).Receive()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/receiver.go:86 +0x75a
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).startResponseRouter()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:223 +0x6d
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC.func1·dwrap·4()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:258 +0x39

Previous write at 0x00c00045a0c0 by goroutine 36:
  github.com/Azure/go-amqp.(*link).mux()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:375 +0x465
  github.com/Azure/go-amqp.attachLink·dwrap·12()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:256 +0x39

Goroutine 37 (running) created at:
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC.func1()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:258 +0x98
  sync.(*Once).doSlow()
      /usr/local/go/src/sync/once.go:68 +0x127
  sync.(*Once).Do()
      /usr/local/go/src/sync/once.go:59 +0x46
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:257 +0xb6
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RetryableRPC.func1()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:190 +0x12c
  github.com/Azure/azure-amqp-common-go/v3.Retry()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/retry.go:41 +0x93
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RetryableRPC()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:186 +0x196
  github.com/Azure/azure-amqp-common-go/v3/cbs.NegotiateClaim()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/cbs/cbs.go:82 +0x6bb
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1.1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:367 +0x226
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:382 +0x144
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:392 +0x26e
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).NegotiateClaim()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:324 +0xdc
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).initWithoutLocking()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:329 +0xc5
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).Get()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:294 +0x312
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Sender).SendMessage()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/sender.go:70 +0x190
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_Queue_Forwarding()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:203 +0xa4b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 36 (running) created at:
  github.com/Azure/go-amqp.attachLink()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:256 +0x194c
  github.com/Azure/go-amqp.(*Session).NewReceiver()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/session.go:95 +0x119
  github.com/Azure/azure-amqp-common-go/v3/rpc.NewLinkWithSession()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:164 +0xa64
  github.com/Azure/azure-amqp-common-go/v3/rpc.NewLink()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:114 +0xb3
  github.com/Azure/azure-amqp-common-go/v3/cbs.NegotiateClaim()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/cbs/cbs.go:54 +0x127
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1.1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:367 +0x226
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:382 +0x144
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:392 +0x26e
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).NegotiateClaim()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:324 +0xdc
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).initWithoutLocking()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:329 +0xc5
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).Get()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:294 +0x312
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Sender).SendMessage()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/sender.go:70 +0x190
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_Queue_Forwarding()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:203 +0xa4b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c0004ba0c0 by goroutine 49:
  github.com/Azure/go-amqp.(*Receiver).Receive()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/receiver.go:86 +0x75a
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).startResponseRouter()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:223 +0x6d
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC.func1·dwrap·4()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:258 +0x39

Previous write at 0x00c0004ba0c0 by goroutine 48:
  github.com/Azure/go-amqp.(*link).mux()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:413 +0x524
  github.com/Azure/go-amqp.attachLink·dwrap·12()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:256 +0x39

Goroutine 49 (running) created at:
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC.func1()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:258 +0x98
  sync.(*Once).doSlow()
      /usr/local/go/src/sync/once.go:68 +0x127
  sync.(*Once).Do()
      /usr/local/go/src/sync/once.go:59 +0x46
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RPC()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:257 +0xb6
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RetryableRPC.func1()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:190 +0x12c
  github.com/Azure/azure-amqp-common-go/v3.Retry()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/retry.go:41 +0x93
  github.com/Azure/azure-amqp-common-go/v3/rpc.(*Link).RetryableRPC()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:186 +0x196
  github.com/Azure/azure-amqp-common-go/v3/cbs.NegotiateClaim()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/cbs/cbs.go:82 +0x6bb
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1.1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:367 +0x226
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:382 +0x144
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:392 +0x26e
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).NegotiateClaim()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:324 +0xdc
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).initWithoutLocking()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:329 +0xc5
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).Get()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:294 +0x312
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:346 +0x10a
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:172 +0x21d
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessage()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:315 +0x53
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_Queue_Forwarding()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:211 +0xae4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 48 (running) created at:
  github.com/Azure/go-amqp.attachLink()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/link.go:256 +0x194c
  github.com/Azure/go-amqp.(*Session).NewReceiver()
      /home/peter/go/pkg/mod/github.com/!azure/go-amqp@v0.16.2/session.go:95 +0x119
  github.com/Azure/azure-amqp-common-go/v3/rpc.NewLinkWithSession()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:164 +0xa64
  github.com/Azure/azure-amqp-common-go/v3/rpc.NewLink()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/rpc/rpc.go:114 +0xb3
  github.com/Azure/azure-amqp-common-go/v3/cbs.NegotiateClaim()
      /home/peter/go/pkg/mod/github.com/!azure/azure-amqp-common-go/v3@v3.2.1/cbs/cbs.go:54 +0x127
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1.1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:367 +0x226
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:382 +0x144
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).startNegotiateClaimRenewer()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:392 +0x26e
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*Namespace).NegotiateClaim()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/namespace.go:324 +0xdc
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).initWithoutLocking()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:329 +0xc5
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal.(*amqpLinks).Get()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/internal/amqpLinks.go:294 +0x312
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:346 +0x10a
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:172 +0x21d
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessage()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/receiver.go:315 +0x53
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_Queue_Forwarding()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:211 +0xae4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
    testing.go:1152: race detected during execution of test
--- FAIL: TestAdminClient_Queue_Forwarding (4.51s)
=== RUN   TestAdminClient_UpdateQueue
--- PASS: TestAdminClient_UpdateQueue (2.46s)
=== RUN   TestAdminClient_GetQueueRuntimeProperties
--- PASS: TestAdminClient_GetQueueRuntimeProperties (3.21s)
=== RUN   TestAdminClient_ListQueues
--- PASS: TestAdminClient_ListQueues (6.21s)
=== RUN   TestAdminClient_ListQueuesRuntimeProperties
--- PASS: TestAdminClient_ListQueuesRuntimeProperties (5.95s)
=== RUN   TestAdminClient_DurationToStringPtr
--- PASS: TestAdminClient_DurationToStringPtr (0.00s)
=== RUN   TestAdminClient_ISO8601StringToDuration
--- PASS: TestAdminClient_ISO8601StringToDuration (0.00s)
=== RUN   TestAdminClient_TopicAndSubscription
--- PASS: TestAdminClient_TopicAndSubscription (5.72s)
=== RUN   TestAdminClient_UpdateTopic
--- PASS: TestAdminClient_UpdateTopic (2.66s)
=== RUN   TestAdminClient_TopicAndSubscriptionRuntimeProperties
--- PASS: TestAdminClient_TopicAndSubscriptionRuntimeProperties (2.46s)
=== RUN   TestAdminClient_ListTopics
==================
WARNING: DATA RACE
Write at 0x00c00032a840 by goroutine 107:
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:688 +0x1d8
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics·dwrap·23()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:693 +0x47

Previous write at 0x00c00032a840 by goroutine 105:
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics.func1()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:688 +0x1d8
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics·dwrap·23()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:693 +0x47

Goroutine 107 (running) created at:
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:686 +0x5f3
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 105 (finished) created at:
  github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.TestAdminClient_ListTopics()
      /home/peter/work/azure-sdk-for-go/sdk/messaging/azservicebus/admin_client_test.go:686 +0x5f3
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
    testing.go:1152: race detected during execution of test
--- FAIL: TestAdminClient_ListTopics (4.51s)
=== RUN   TestAdminClient_ListTopicsRuntimeProperties
--- PASS: TestAdminClient_ListTopicsRuntimeProperties (5.46s)
=== RUN   TestAdminClient_ListSubscriptions
--- PASS: TestAdminClient_ListSubscriptions (2.26s)
=== RUN   TestAdminClient_ListSubscriptionRuntimeProperties
--- PASS: TestAdminClient_ListSubscriptionRuntimeProperties (2.01s)
=== RUN   TestAdminClient_UpdateSubscription
--- PASS: TestAdminClient_UpdateSubscription (4.39s)
=== RUN   TestAdminClient_LackPermissions_Queue
--- PASS: TestAdminClient_LackPermissions_Queue (5.16s)
=== RUN   TestAdminClient_LackPermissions_Topic
--- PASS: TestAdminClient_LackPermissions_Topic (5.69s)
=== RUN   TestAdminClient_LackPermissions_Subscription
--- PASS: TestAdminClient_LackPermissions_Subscription (9.42s)
=== RUN   TestNewClientWithAzureIdentity
    client_test.go:26: Azure Identity compatible credentials not configured
--- SKIP: TestNewClientWithAzureIdentity (1.84s)
=== RUN   TestNewClientUnitTests
=== RUN   TestNewClientUnitTests/WithTokenCredential
=== RUN   TestNewClientUnitTests/CloseAndLinkTracking
--- PASS: TestNewClientUnitTests (0.00s)
    --- PASS: TestNewClientUnitTests/WithTokenCredential (0.00s)
    --- PASS: TestNewClientUnitTests/CloseAndLinkTracking (0.00s)
=== RUN   TestErrClosed
--- PASS: TestErrClosed (0.00s)
=== RUN   TestMessageBatchUnitTests
=== RUN   TestMessageBatchUnitTests/addMessages
=== RUN   TestMessageBatchUnitTests/addTooManyMessages
--- PASS: TestMessageBatchUnitTests (0.00s)
    --- PASS: TestMessageBatchUnitTests/addMessages (0.00s)
    --- PASS: TestMessageBatchUnitTests/addTooManyMessages (0.00s)
=== RUN   TestMessageSettlementUsingReceiver
--- PASS: TestMessageSettlementUsingReceiver (4.09s)
=== RUN   TestDeferredMessages
=== RUN   TestDeferredMessages/Abandon
    messageSettler_test.go:76: This test is currently broken, https://github.com/Azure/azure-sdk-for-go/issues/15626
=== RUN   TestDeferredMessages/Complete
=== RUN   TestDeferredMessages/Defer
--- PASS: TestDeferredMessages (3.48s)
    --- SKIP: TestDeferredMessages/Abandon (0.00s)
    --- PASS: TestDeferredMessages/Complete (1.57s)
    --- PASS: TestDeferredMessages/Defer (0.29s)
=== RUN   TestDeferredMessage_DeadLettering
--- PASS: TestDeferredMessage_DeadLettering (4.41s)
=== RUN   TestMessageSettlementUsingOnlyBackupSettlement
--- PASS: TestMessageSettlementUsingOnlyBackupSettlement (3.89s)
=== RUN   TestMessageSettlementWithDeferral
--- PASS: TestMessageSettlementWithDeferral (1.37s)
=== RUN   TestMessageUnitTest
=== RUN   TestMessageUnitTest/toAMQPMessage
--- PASS: TestMessageUnitTest (0.00s)
    --- PASS: TestMessageUnitTest/toAMQPMessage (0.00s)
=== RUN   TestAMQPMessageToReceivedMessage
=== RUN   TestAMQPMessageToReceivedMessage/empty_message
=== RUN   TestAMQPMessageToReceivedMessage/annotations
--- PASS: TestAMQPMessageToReceivedMessage (0.00s)
    --- PASS: TestAMQPMessageToReceivedMessage/empty_message (0.00s)
    --- PASS: TestAMQPMessageToReceivedMessage/annotations (0.00s)
=== RUN   TestAMQPMessageToMessage
--- PASS: TestAMQPMessageToMessage (0.00s)
=== RUN   TestProcessorReceiveWithDefaults
--- PASS: TestProcessorReceiveWithDefaults (9.18s)
=== RUN   TestProcessorReceiveWith100MessagesWithMaxConcurrency
--- PASS: TestProcessorReceiveWith100MessagesWithMaxConcurrency (3.43s)
=== RUN   TestProcessorUnitTests
--- PASS: TestProcessorUnitTests (0.00s)
=== RUN   TestReceiverSendFiveReceiveFive
--- PASS: TestReceiverSendFiveReceiveFive (3.53s)
=== RUN   TestReceiverForceTimeoutWithTooFewMessages
--- PASS: TestReceiverForceTimeoutWithTooFewMessages (4.09s)
=== RUN   TestReceiverAbandon
--- PASS: TestReceiverAbandon (3.37s)
=== RUN   TestReceiveWithEarlyFirstMessageTimeout
--- PASS: TestReceiveWithEarlyFirstMessageTimeout (3.20s)
=== RUN   TestReceiverSendAndReceiveManyTimes
--- PASS: TestReceiverSendAndReceiveManyTimes (18.64s)
=== RUN   TestReceiverDeferAndReceiveDeferredMessages
--- PASS: TestReceiverDeferAndReceiveDeferredMessages (3.28s)
=== RUN   TestReceiverPeek
--- PASS: TestReceiverPeek (3.28s)
=== RUN   TestReceiver_RenewMessageLock
--- PASS: TestReceiver_RenewMessageLock (9.20s)
=== RUN   TestReceiverOptions
--- PASS: TestReceiverOptions (0.00s)
=== RUN   TestReceiverDeferUnitTests
--- PASS: TestReceiverDeferUnitTests (0.00s)
=== RUN   Test_Sender_SendBatchOfTwo
--- PASS: Test_Sender_SendBatchOfTwo (2.88s)
=== RUN   Test_Sender_UsingPartitionedQueue
--- PASS: Test_Sender_UsingPartitionedQueue (3.28s)
=== RUN   Test_Sender_SendMessages
--- PASS: Test_Sender_SendMessages (3.27s)
=== RUN   Test_Sender_SendMessages_resend
--- PASS: Test_Sender_SendMessages_resend (3.88s)
=== RUN   Test_Sender_ScheduleMessages
--- PASS: Test_Sender_ScheduleMessages (21.73s)
=== RUN   TestSessionReceiver_acceptSession
--- PASS: TestSessionReceiver_acceptSession (3.48s)
=== RUN   TestSessionReceiver_blankSessionIDs
    session_receiver_test.go:58: Can't run blank session ID test because of issue
--- SKIP: TestSessionReceiver_blankSessionIDs (0.00s)
=== RUN   TestSessionReceiver_acceptSessionButAlreadyLocked
--- PASS: TestSessionReceiver_acceptSessionButAlreadyLocked (3.70s)
=== RUN   TestSessionReceiver_acceptNextSession
--- PASS: TestSessionReceiver_acceptNextSession (3.25s)
=== RUN   TestSessionReceiver_noSessionsAvailable
    session_receiver_test.go:152: Really slow test (since it has to wait for a timeout from the service)
--- SKIP: TestSessionReceiver_noSessionsAvailable (0.00s)
=== RUN   TestSessionReceiver_nonSessionReceiver
--- PASS: TestSessionReceiver_nonSessionReceiver (3.91s)
=== RUN   TestSessionReceiver_RenewSessionLock
--- PASS: TestSessionReceiver_RenewSessionLock (28.45s)
=== RUN   Test_toReceiverOptions
--- PASS: Test_toReceiverOptions (0.00s)
=== CONT  
    testing.go:1152: race detected during execution of test
FAIL
exit status 1
FAIL    github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus    235.249s
``` 

Even if it is just in the close method, this bug means we cannot run our integration tests using ASB with the race detector enabled.

--- a/sdk/messaging/azservicebus/CHANGELOG.md
+++ b/sdk/messaging/azservicebus/CHANGELOG.md
@@ -1,5 +1,11 @@
 # Release History
 
+## 0.3.1 (2021-11-16)
+
+### Bugs Fixed
+
+- Updating go-amqp to v0.16.4 to fix a race condition found when running `go test -race`.  Thanks to @peterzeller for reporting this issue. PR: #16168
+
 ## 0.3.0 (2021-11-12)
 
 ### Features Added

--- a/sdk/messaging/azservicebus/go.mod
+++ b/sdk/messaging/azservicebus/go.mod
@@ -8,7 +8,7 @@ require (
 	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.20.0
 	github.com/Azure/azure-sdk-for-go/sdk/azidentity v0.12.0
 	github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.1
-	github.com/Azure/go-amqp v0.16.3
+	github.com/Azure/go-amqp v0.16.4
 	github.com/Azure/go-autorest/autorest v0.11.18
 	github.com/Azure/go-autorest/autorest/adal v0.9.13
 	github.com/Azure/go-autorest/autorest/date v0.3.0

--- a/sdk/messaging/azservicebus/internal/constants.go
+++ b/sdk/messaging/azservicebus/internal/constants.go
@@ -19,4 +19,4 @@ const (
 // TODO: this should move into a proper file. Need to resolve some interdependency
 // issues between the public and internal packages first.
 // Version is the semantic version number
-const Version = "v0.3.0"
+const Version = "v0.3.1"

